File: epi_cpp/Arrays/18_random_permutation.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <random>

using namespace std;

void RandomSampling(vector<int>* A_ptr, int k){
    vector<int>& A = *A_ptr;
    default_random_engine seed((random_device())());
    for(int i = 0; i < k; i++){
        swap(A[i], A[uniform_int_distribution<int>{i, static_cast<int>(A.size()-1)}(seed)]);
    }
}

vector<int> ComputeRandomPermutation(int n){
    vector<int> permutation(n);
    iota(permutation.begin(), permutation.end(), 0);
    RandomSampling(&permutation, permutation.size());
    return permutation;
}

int main(){
    int n = 5;
    vector<int> permutation;
    permutation = ComputeRandomPermutation(n);
    for(auto val : permutation){
        cout << val << " ";
    }
    cout << endl;
    return 0;
}

File: epi_python/02_Arrays/18_random_permutation.py
from typing import List
import random

def random_sampling(k: int, A: List[int]) -> None:
    for i in range(k):
        r = random.randint(i, len(A) - 1)
        A[i], A[r] = A[r], A[i]

def compute_random_permutation(n: int) -> List[int]:
    permutation = list(range(n))
    random_sampling(n, permutation)
    return permutation

if __name__ == '__main__':
    n = 5
    perm = compute_random_permutation(n)
    print(*perm, end="\n")